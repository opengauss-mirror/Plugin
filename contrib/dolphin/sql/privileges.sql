create role lowuser password 'lowuser@123';

create schema test_privileges;
create table test_privileges.t1 (c1 int);

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');

alter default privileges in schema test_privileges grant SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables to public;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
create table test_privileges.t2 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');

alter default privileges in schema test_privileges revoke SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables from public;
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');
create table test_privileges.t3 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'VACUUM');

alter default privileges in schema test_privileges revoke grant option for SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables from public;
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');
create table test_privileges.t4 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'VACUUM');

GRANT SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 to lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
REVOKE GRANT OPTION FOR SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 from lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
REVOKE SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 from lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');

drop user lowuser;
drop schema test_privileges cascade;
