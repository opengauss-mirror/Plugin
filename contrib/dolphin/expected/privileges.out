create role lowuser password 'lowuser@123';
create schema test_privileges;
create table test_privileges.t1 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
 has_table_privilege 
---------------------
 f
(1 row)

alter default privileges in schema test_privileges grant SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables to public;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
 has_table_privilege 
---------------------
 f
(1 row)

create table test_privileges.t2 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');
 has_table_privilege 
---------------------
 t
(1 row)

alter default privileges in schema test_privileges revoke SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables from public;
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');
 has_table_privilege 
---------------------
 t
(1 row)

create table test_privileges.t3 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'SELECT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'INSERT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'UPDATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'DELETE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'TRUNCATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'REFERENCES');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'ALTER');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'DROP');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'COMMENT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'INDEX');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t3', 'VACUUM');
 has_table_privilege 
---------------------
 f
(1 row)

alter default privileges in schema test_privileges revoke grant option for SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on tables from public;
SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'SELECT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INSERT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'UPDATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DELETE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'TRUNCATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'REFERENCES');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'ALTER');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'DROP');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'COMMENT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'INDEX');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t2', 'VACUUM');
 has_table_privilege 
---------------------
 t
(1 row)

create table test_privileges.t4 (c1 int);
SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'SELECT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'INSERT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'UPDATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'DELETE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'TRUNCATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'REFERENCES');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'ALTER');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'DROP');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'COMMENT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'INDEX');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t4', 'VACUUM');
 has_table_privilege 
---------------------
 f
(1 row)

GRANT SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 to lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
 has_table_privilege 
---------------------
 t
(1 row)

REVOKE GRANT OPTION FOR SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 from lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
 has_table_privilege 
---------------------
 t
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
 has_table_privilege 
---------------------
 t
(1 row)

REVOKE SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, ALTER, DROP, COMMENT, INDEX, VACUUM on test_privileges.t1 from lowuser;
SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'SELECT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INSERT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'UPDATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DELETE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'TRUNCATE');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'REFERENCES');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'ALTER');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'DROP');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'COMMENT');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'INDEX');
 has_table_privilege 
---------------------
 f
(1 row)

SELECT has_table_privilege('lowuser', 'test_privileges.t1', 'VACUUM');
 has_table_privilege 
---------------------
 f
(1 row)

drop user lowuser;
drop schema test_privileges cascade;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to table test_privileges.t1
drop cascades to table test_privileges.t2
drop cascades to table test_privileges.t3
drop cascades to table test_privileges.t4
